consensus/src/fixed/orchestrator.rs

```diff
@@ -273,17 +242,32 @@ impl<E: Clock + Rng + Spawner, A: Application> Orchestrator<E, A> {
         }
 
         // Record what we learned
-        match self.knowledge.entry(proposal.height) {
-            Entry::Vacant(e) => {
-                if proposal.height > self.last_finalized {
+        if proposal.height > self.last_finalized {
+            // Add to notarized if not finalized
+            match self.knowledge.get_mut(&proposal.height) {
+                Some(Knowledge::Notarized(seen)) => {
+                    seen.insert(proposal.view, hash.clone());
+                }
+                None => {
                     let mut seen = BTreeMap::new();
                     seen.insert(proposal.view, hash.clone());
-                    e.insert(Knowledge::Notarized(seen));
-                } else {
-                    e.insert(Knowledge::Finalized(hash.clone()));
+                    self.knowledge
+                        .insert(proposal.height, Knowledge::Notarized(seen));
                 }
+                _ => {}
+            }
+        } else {
+            // TODO: clean this up
+
+            // Insert as final (in case it doesn't exist)
+            let mut start = (proposal.height, hash.clone());
+            if let Some(Knowledge::Finalized(_)) = self.knowledge.get(&proposal.height) {
+                debug!("overriding backfill start to parent");
+                start = (proposal.height - 1, proposal.parent.clone());
             }
-            Entry::Occupied(_) => {}
+
+            // Finalize this block and all blocks we have that are ancestors of this block
+            self.backfill_finalization(start.0, start.1);
```

Comment:
This was a bug (unrelated to this change). Prior to this we could've not bubbled up our info about finalization prior to invoking `notify` (leading to skips in finalization messages).